### PR TITLE
Added Component Instance

### PR DESCRIPTION
**Pull request by petrochuk**

### CODE REVIEW

File: src/Persistence/Templates/BuiltInTemplates.cs [hunk 1]
```diff
@@ -3,15 +3,20 @@
...
 
-public static class BuiltInTemplates
+public class BuiltInTemplates
 {
-    public const string App = "Appinfo";
-    public const string Host = "HostControl";
-    public const string Screen = "Screen";
-    public const string Component = "Component";
-    public const string CommandComponent = "CommandComponent";
+    public static readonly (string Name, string Id) App = ("Appinfo", "http://microsoft.com/appmagic/appinfo");
```
Comment (mizrael) [inline src/Persistence/Templates/BuiltInTemplates.cs:8]: nit: we should evaluate a `TemplateKey` struct with `Name` and `Id`. Tuples are not very manageable.

File: src/Persistence/Templates/ControlFactory.cs [hunk 2]
```diff
@@ -17,11 +17,15 @@ public ControlFactory(IControlTemplateStore controlTemplateStore)
         _controlTemplateStore = controlTemplateStore ?? throw new ArgumentNullException(nameof(controlTemplateStore));
     }
 
-    public Control Create(string name, string template, string? variant = null, ControlPropertiesCollection? properties = null, IList<Control>? children = null)
+    public Control Create(string name, string template, string? componentInstance = null, string? variant = null,
```
Comment (mizrael) [inline src/Persistence/Templates/ControlFactory.cs:20]: `componentInstance` seems to be the `componentDefinitionName`, is that right?

File: src/Persistence/Models/ComponentDefinition.cs [hunk 3]
```diff
@@ -12,28 +11,29 @@ namespace Microsoft.PowerPlatform.PowerApps.Persistence.Models;
 /// <summary>
 /// This class is meant to describe a Component definition.
 /// </summary>
-[FirstClass(templateName: BuiltInTemplates.Component)]
```
Comment (mizrael) [inline src/Persistence/Models/ComponentDefinition.cs:15]: why we're not considering this a First Class control type?

File: src/Persistence/Models/ComponentInstance.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+using YamlDotNet.Serialization;
+
+namespace Microsoft.PowerPlatform.PowerApps.Persistence.Models;
+
+public record ComponentInstance : Control
+{
+    /// <summary>
+    /// The template id for a local component instance.
+    /// </summary>
+    public const string ComponentInstanceTemplateId = "http://localhost/Component";
```
[the remark this answers is not in the document]
Comment (petrochuk) [reply src/Persistence/Models/ComponentInstance.cs:15]: Can you elaborate why is this redundant? Component instances have local templates

File: src/Persistence/Templates/ControlFactory.cs (hunk 2 above)
Comment (petrochuk) [reply src/Persistence/Templates/ControlFactory.cs:20]: Good point. I'll rename to make it clearer

File: src/Persistence/Models/ComponentDefinition.cs (hunk 3 above)
Comment (petrochuk) [reply src/Persistence/Models/ComponentDefinition.cs:15]: Because they have custom Yaml converters

**Review by mizrael — APPROVED** [review]

**Review by shshetty-microsoft — APPROVED** [review]

**Pull request merged**